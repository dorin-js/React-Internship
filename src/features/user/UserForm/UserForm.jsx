/* eslint-disable no-useless-escape */
import React from 'react';
import { PropTypes } from 'prop-types';
import { useForm, FormProvider } from 'react-hook-form';
import { useCreateNewUserMutation, useUpdateUserMutation } from '../../../services/api/apiService';
import { Button } from '../../../common/components/Button';
import { Snackbar } from '../../../common/components/Snackbar';
import FormInput from '../../../common/components/FormInput/FormInput';
import DatePicker from '../../../common/components/DatePicker/DatePicker';

const UserForm = ({ isEditing, user }) => {
  const {
    firstname, lastname, email, birthDate, id,
  } = user;

  const methods = useForm(
    isEditing && {
      defaultValues: {
        firstname,
        lastname,
        email,
        birthDate,
        id,
      },
    },
  );

  const [updateUser, {
    isLoading: isUpdatingInProgress,
    isError: isUpdateUserError,
    error: updateUserError,
  }] = useUpdateUserMutation();

  const [createNewUser, {
    isLoading: isCreatingInProgress,
    isError: isCreateNewUserError,
    error: createNewUserError,
  }] = useCreateNewUserMutation();

  const onSubmit = (data) => {
    if (isEditing) {
      updateUser(data);
    } else {
      createNewUser(data);
    }
  };

  return (
    <>
      <FormProvider {...methods}>
        <form
          onSubmit={methods.handleSubmit(onSubmit)}
          name="form"
        >
          <FormInput
            name="firstname"
            placeholder="First Name"
          />
          <FormInput
            name="lastname"
            placeholder="Last Name"
          />
          <FormInput
            name="email"
            placeholder="Email"
            data-testid="email-input-field"
          />
          <DatePicker
            name="birthDate"
            type="date"
            placeholder="Date of birth"
            min="1900-01-01"
          />
          <Button
            type="submit"
            aria-label={isEditing ? 'Update user' : 'Create user'}
            value={(isUpdatingInProgress || isCreatingInProgress) ? 'Saving...' : 'Save'}
          />
        </form>
      </FormProvider>
      {
        (isCreateNewUserError || isUpdateUserError)
        && (
          <Snackbar
            type="error"
            timeout={4000}
            message={
              createNewUserError?.data?.error
              || createNewUserError?.error
              || updateUserError?.data?.error
              || updateUserError?.error
            }
          />
        )
      }
    </>
  );
};

UserForm.propTypes = {
  user: PropTypes.PropTypes.shape({
    id: PropTypes.string,
    firstname: PropTypes.string,
    lastname: PropTypes.string,
    email: PropTypes.string,
    birthDate: PropTypes.string,
  }),
  isEditing: PropTypes.bool,
};
UserForm.defaultProps = {
  user: {},
  isEditing: undefined,
};

export default UserForm;
