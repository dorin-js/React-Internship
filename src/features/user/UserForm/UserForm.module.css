.createUserForm input {
  display: block;
  padding: 4px;
  width: 100%;
  max-width: 100%;
  border: 1px solid #24292e;
  border-radius: 2px;
}

.createUserForm p {
  margin-top: 5px;
  padding: 0;
  color: rgb(255, 66, 66);
  font-size: 0.9em;
}

.inputContainer {
  margin: 25px auto;
}