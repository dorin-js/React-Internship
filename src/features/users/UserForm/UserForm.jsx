import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classes from './UserForm.module.css';
import { usersApi } from '../../../common/services/usersApi/usersApi';
import { Snackbar } from '../../../common/components/Snackbar';
import { Button } from '../../../common/components/Button';

const defaultFormData = {
  name: '',
  lastname: '',
  email: '',
  birth: '',
};

const UserForm = ({ onCreateUser = undefined }) => {
  const [form, setForm] = useState(defaultFormData);
  const [loading, setLoading] = useState();
  const [error, setError] = useState(null);

  const createUser = async (body) => {
    setLoading(true);
    try {
      const res = await usersApi.postUser(body);
      if (res) {
        onCreateUser();
      }
      // onCreateUser();
    } catch (e) {
      setError(e.message);
    }
    setLoading(false);
  };

  const onValueChanged = (value) => {
    setForm((prevState) => ({ ...prevState, ...value }));
  };

  return (
    <>
      <form
        name="form"
        className={classes.createUserForm}
      >
        <input
          required
          placeholder="First Name"
          value={form.name}
          onChange={(e) => onValueChanged({ name: e.target.value })}
        />
        <input
          required
          placeholder="Last Name"
          value={form.lastname}
          onChange={(e) => onValueChanged({ lastname: e.target.value })}
        />
        <input
          type="email"
          data-testid="email-input-field"
          required
          placeholder="Email"
          value={form.email}
          onChange={(e) => onValueChanged({ email: e.target.value })}
        />
        <input
          type="date"
          required
          placeholder="Date of birth"
          value={form.birth}
          onChange={(e) => onValueChanged({ birth: e.target.value })}
        />
        <div className="buttonsContainer">
          <Button data-testid="submit-button" onClick={() => createUser(form)} value={loading ? 'Creating...' : 'Create User'} />
        </div>
      </form>
      {
        error && <Snackbar type="error" message={error} timeout={4} />
      }
    </>
  );
};

UserForm.propTypes = {
  onCreateUser: PropTypes.func,
};
UserForm.defaultProps = {
  onCreateUser: () => undefined,
};

export default UserForm;
