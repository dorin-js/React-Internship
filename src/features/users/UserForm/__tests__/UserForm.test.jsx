import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import UserForm from '../UserForm';

const mockPostUser = jest.fn();
jest.mock('../../../../common/services/usersApi', () => jest.fn().mockImplementation(() => ({ postUser: mockPostUser })));

describe('Render form', () => {
  it.skip('should render correctly', () => {
    const { baseElement } = render(<UserForm onCreateUser={mockPostUser} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('creates new user', async () => {
    // const mockSubmit = jest.fn();
    const onCreateUser = (items) => {
      console.log(items);
    };
    const {
      getByPlaceholderText, getByTestId,
    } = render(<UserForm onCreateUser={onCreateUser} />);

    fireEvent.change(getByPlaceholderText('First Name'), { target: { value: 'Test' } });
    fireEvent.change(getByPlaceholderText('Last Name'), { target: { value: 'Test1' } });
    fireEvent.change(getByPlaceholderText('Email'), { target: { value: 'test@mail.co' } });
    fireEvent.change(getByPlaceholderText('Date of birth'), { target: { value: '27-10-2001' } });

    fireEvent.click(getByTestId('submit-button'));
    await waitFor(async () => {
      expect(mockPostUser).toHaveBeenCalled();
    });
  });
});
