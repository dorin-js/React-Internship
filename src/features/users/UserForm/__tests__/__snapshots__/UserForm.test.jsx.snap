// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render form should render correctly 1`] = `
<body>
  <div>
    <form
      class="createUserForm"
      name="form"
    >
      <input
        placeholder="First Name"
        required=""
        value=""
      />
      <input
        placeholder="Last Name"
        required=""
        value=""
      />
      <input
        data-testid="email-input-field"
        placeholder="Email"
        required=""
        type="email"
        value=""
      />
      <input
        placeholder="Date of birth"
        required=""
        type="date"
        value=""
      />
      <div
        class="buttonsContainer"
      >
        <button
          class="button"
          data-testid="submit-button"
          name="Create User"
          type="button"
        >
          <span
            class="buttonText"
          >
            Create User
          </span>
        </button>
      </div>
    </form>
  </div>
</body>
`;
